### PR TITLE
SWIFT-1143 Test proxy as mongos

### PR DESCRIPTION
**Pull request by patrickfreed**
SWIFT-1143

This PR adds testing for the Atlas proxy as a mongos. As part of that, it also adds support for the unified format CRUD tests (SWIFT-1183). SWIFT-762 is also completed as part of this.

### CODE REVIEW

**Review by patrickfreed — COMMENTED** [review]
The failures I believe are due to the versioned API changes which should be fixed as part of #638.

File: Tests/MongoSwiftTests/AsyncTestUtils.swift [hunk 1]
```diff
@@ -27,6 +28,16 @@ extension MongoClient {
                 opts.serverAPI!.version = apiVersion
             }
         }
+
```
Comment (kmahar) [inline Tests/MongoSwiftTests/AsyncTestUtils.swift:31]: all the logic we're starting to repeat across the sync/async `makeTestClient` methods makes me think we should have some kind of common method to resolve client options and environment variables in `TestsCommon`? seems out of scope for this PR though

**Review by kmahar — COMMENTED** [review]
looks great overall! just have some various small questions and comments

File: Tests/MongoSwiftTests/AsyncTestUtils.swift (hunk 1 above)
Comment (patrickfreed) [reply Tests/MongoSwiftTests/AsyncTestUtils.swift:31]: Yeah that's definitely a good idea. I ran into a bunch of bugs over the course of implementing this due to differences between the two implementations. Filed SWIFT-1241 to cover the work for doing this.

File: Tests/MongoSwiftTests/ChangeStreamTests.swift [hunk 2]
```diff
@@ -54,7 +54,12 @@ final class ChangeStreamTests: MongoSwiftTestCase {
             )
             let unmetRequirement = try client.getUnmetRequirement(testRequirements)
             guard unmetRequirement == nil else {
-                printSkipMessage(testName: self.name, unmetRequirement: unmetRequirement!)
+                switch unmetRequirement {
+                case .minServerVersion, .maxServerVersion:
+                    print("Skipping test; see SWIFT-722")
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Tests/MongoSwiftTests/ChangeStreamTests.swift]: also removed.

File: Sources/TestsCommon/CommonTestUtils.swift [hunk 3]
```diff
@@ -263,14 +286,28 @@ public struct TestRequirement: Decodable {
         if let topologies = self.topologies {
             // When matching a "sharded" topology, test runners MUST accept any type of sharded cluster (i.e. "sharded"
             // implies "sharded-replicaset", but not vice versa).
-            if topology == .shardedReplicaSet && topologies.contains(.sharded) {
```
[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/TestsCommon/CommonTestUtils.swift:268]: since mongo-orchestration always sets up a `.shardedReplicaSet` (see [configs](https://github.com/mongodb-labs/drivers-evergreen-tools/tree/master/.evergreen/orchestration/configs/sharded_clusters)) I think this evaluates to true basically anytime we test against a sharded cluster and `sharded` is among the topologies to run on.... I'm guessing we've maybe just gotten lucky and none of the tests have had extra requirements beyond that? we don't actually run many tests with the unified runner yet

File: Tests/MongoSwiftSyncTests/CrudTests.swift [hunk 4]
```diff
@@ -32,10 +32,6 @@ final class CrudTests: MongoSwiftTestCase {
 
             // For each file, execute the test cases contained in it
             for (i, test) in try file.makeTests().enumerated() {
-                if type(of: test) == CountTest.self {
```
[the remark this answers is not in the document]
Comment (kmahar) [reply Tests/MongoSwiftSyncTests/CrudTests.swift:35]: yeah I think that would make sense

**Review by kmahar — APPROVED** [review]
one comment but I don't need to re-review. thanks for the explanations and nice job with this! LGTM

File: Tests/MongoSwiftSyncTests/CrudTests.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Tests/MongoSwiftSyncTests/CrudTests.swift:35]: fixed

**Comment by patrickfreed** [comment]
Going to do one more CI run to verify it's all still green.
